### PR TITLE
[ENH] Handling of SurfaceData in compute_contrast

### LINKED ISSUE #4020: [ENH] Surfaced-based GLM needs the function "nilearn.glm.compute_fixed_effects" to accept 1-D array contrast and variance
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe your proposed enhancement in detail.

I was trying to do surface-based GLM on my data. But my data have several sessions, and there is no function available to compute the fixed effects given the input contrast is a 1-D numpy.array instead of Nifti1Images. Your need to have an function to do that with accepting input other than Nifti1Images when surface-based analysis is implemented.

### Benefits to the change

Your example of surface-based GLM is only for one session data. Your functions for surface-based analysis cannot deal with data with more than one session when there is need to compute fixed effects.

### Pseudocode for the new behavior, if applicable

```python
# insert your code below

```

### PR DESCRIPTION
**Pull request by bthirion**
- Closes #4020

### CODE REVIEW

**Comment by bthirion** [comment]
Happy to get reviews on that one.

File: nilearn/glm/contrasts.py [hunk 1]
```diff
@@ -525,15 +527,30 @@ def compute_fixed_effects(
...
     elif mask is None:
-        masker = NiftiMasker().fit(contrast_imgs)
+        if isinstance(contrast_imgs[0], SurfaceImage):
+            surf_data = {}
+            for part in contrast_imgs[0].mesh.parts:
+                surf_data[part] = np.ones(
+                    contrast_imgs[0].data.parts[part].shape[0], dtype=bool
+                )
+            mask_img = SurfaceImage(mesh=contrast_imgs[0].mesh, data=surf_data)
+            masker = SurfaceMasker(mask_img=mask_img).fit(contrast_imgs[0])
```
Comment (Remi-Gau) [inline nilearn/glm/contrasts.py:540]: Can we not jsut do ?

```
            masker = SurfaceMasker().fit(contrast_imgs[0])
```

the fit of the surface masker does implicitly build a mask if none is passed at instanciation:

https://github.com/nilearn/nilearn/blob/e2c7e3c203f22e1cc40d5678e94dc4b3421c7101/nilearn/experimental/surface/maskers.py#L152

Comment (bthirion) [reply nilearn/glm/contrasts.py:540]: I don't understand: in the case `mask_img` is provided, we should actually use it.

Comment (Remi-Gau) [reply nilearn/glm/contrasts.py:540]: agreed
my comment was unclear: let me try again

File: nilearn/glm/contrasts.py [hunk 2]
```diff
@@ -525,15 +527,30 @@ def compute_fixed_effects(
...
+        else:
+            masker = NiftiMasker().fit(contrast_imgs)
+    elif isinstance(mask, SurfaceImage):
+        masker = SurfaceMasker().fit(contrast_imgs[0])
```
Comment (Remi-Gau) [inline nilearn/glm/contrasts.py:544]:
```suggestion
        if isinstance(contrast_imgs[0], SurfaceImage):
            masker = SurfaceMasker().fit(contrast_imgs[0])
        else:
            masker = NiftiMasker().fit(contrast_imgs)
    elif isinstance(mask, SurfaceImage):
        masker = SurfaceMasker(mask_img=mask).fit(contrast_imgs[0])
```

Given how the surface masker works (estimates mask from images passed at fit time it no mask was provided at instantiation) I think we can simplify this way.

Does it make sense?

File: nilearn/glm/tests/test_first_level.py [hunk 3]
```diff
@@ -2273,3 +2273,25 @@ def test_first_level_from_bids_subject_order_with_labels(tmp_path):
...
+    """Smoke test of compute_fixed_effects with surface data."""
+    mini_img, _ = _make_surface_glm_data(5)
+    masker = SurfaceMasker().fit(mini_img)
+    model = FirstLevelModel(mask_img=masker, t_r=2.0)
+    events = basic_paradigm()
+    model.fit([mini_img, mini_img], events=[events, events])
+    result = model.compute_contrast("c0", output_type="all")
+    effect = result["effect_size"]
+    variance = result["effect_variance"]
+    for mask in [SurfaceMasker(mask_img=masker.mask_img_), None]:
```
Comment (Remi-Gau) [inline nilearn/glm/tests/test_first_level.py:2288]: refactor this in a follow up PR to use test parametrization

**Comment by bthirion** [comment]
Should be ready for another review.

**Comment by Remi-Gau** [comment]
FYI: I fixed style in 786ce208ad46e9f5085f3bef1b0fb3da5a6c9f7c

**Review by Remi-Gau — APPROVED** [review]
Waiting for CI to be done but good with me.
